### ISSUE TITLE
Flexible work Shop with Capacity

### ISSUE THREAD

**Issue by YooTimmy** [google/or-tools#1584]
Hello all,

I am trying to add capacity constrain for the FJP problem. However during the implementation, I notice that the solve would get stuck when I increase the number of jobs just a little bit (say from 9 to 15).. The code I have tried and more details could be found from this [link](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/or-tools-discuss/0syUImixcFI).
Could anyone give me some suggestions how to improve or change my current code to resolve this issue? Many thanks!

**Comment by lperron** [comment]
Can you send the code with 15 jobs ?
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by lperron** [comment]
Actually, let me rephrase


job_profile_1 = [[(3, 0), (2, 2)], [(2, 1)]]
job_profile_2 = [[(4, 0), (3, 2)], [(3, 1)]]
job_profile_list = [job_profile_1, job_profile_2]
# The total amount of tasks to be planned on the machine

# 2 orders:
orders = [[(0, 3), (1, 3)], [(1, 3)]]

I understand the job profiles.
I do not understand the orders, do we create 3 copies of job 1 and 3 + 3
copies of job 2, or do we create 3 jobs that will require a capacity of 3
on the machines ?
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by YooTimmy** [comment]
Hello Laurent,

Thanks for getting back. Short answer your first understanding is correct.

For Orders, it is more like simulating in real scenarios, where customer may need different jobs in each order. Let me explain the orders =  [[(0, 3), (1, 3)], [(1, 3)]]

Here there are 2 orders. First order is [(0, 3), (1, 3)], which means we create 3 copies of job 1 (0,3) and 3 copies of job 2 (1,3). Second Order is [(1,3)], which just means we need to create 3 copies of job 2. In other words, here we have a total 9 jobs that need to be planned (3 copies of job 1, and 6 copies of job 2). In this case, I can able to get a valid schedule.

When I mentioned changing 9 jobs to 15 jobs, I was just change the orders to [[(0, 5), (1, 5)], [(1, 5)]]. In this case seems the solver will stuck on my PC.

**Comment by lperron** [comment]
OK. So I believe that quite fast, you get the optimal solution.
Proving optimality is hard, and I am looking at redundant constraint for it.

I have not found anything yet.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by YooTimmy** [comment]
Hello Laurent,

Thanks for the feedback.

> So I believe that quite fast, you get the optimal solution. 

I am not sure if I capture you correctly. For now in the case of 15 jobs, I have tried to let it run for like 10 mins, seems it is not solved by then. Maybe I should wait some more time?
![image](https://user-images.githubusercontent.com/32185776/65133114-b2892e00-da34-11e9-9562-5d23510dc281.png)

If that is the case, is the reason because perhaps adding the capacity consideration to flexible job shop problem would increase the complexity significantly? I thought 15 jobs is not that much but seems not the case then.

**Comment by lperron** [comment]
I believe 17 is optimal. So you get it in 0.04s.
Proof is just hard.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by YooTimmy** [comment]
Got it!

Would it be possible to let solver return the variable values when it is not able to prove it is optimal say in 60s? I guess that will be helpful as well as a potential optimal solution.

**Comment by YooTimmy** [comment]
Thank you Laurent, now I understand your point.

**Comment by lperron** [comment]
Just add a time limit.

https://github.com/google/or-tools/blob/stable/ortools/sat/doc/solver.md#changing-the-parameters-of-the-solver
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00
